crates/driver/src/domain/competition/sorting.rs

```diff
@@ -0,0 +1,122 @@
+use {
+    crate::{
+        domain::{
+            competition::{auction::Tokens, order},
+            eth,
+        },
+        util,
+    },
+    chrono::{Duration, Utc},
+    std::{cmp::Ordering, sync::Arc},
+};
+
+pub trait OrderComparator: Send + Sync {
+    fn compare(
+        &self,
+        order_a: &order::Order,
+        order_b: &order::Order,
+        tokens: &Tokens,
+        solver: &eth::H160,
+    ) -> Ordering;
+}
+
+impl<F> OrderComparator for F
+where
+    F: Fn(&order::Order, &order::Order, &Tokens, &eth::H160) -> Ordering + Send + Sync,
+{
+    fn compare(
+        &self,
+        order_a: &order::Order,
+        order_b: &order::Order,
+        tokens: &Tokens,
+        solver: &eth::H160,
+    ) -> Ordering {
+        self(order_a, order_b, tokens, solver)
+    }
+}
+
+pub trait OrderingKey: Send + Sync + 'static {
+    type Key: Ord + Send + Sync + 'static;
+
+    fn key(&self, order: &order::Order, tokens: &Tokens, solver: &eth::H160) -> Self::Key;
+
+    /// Returns a comparator that compares two orders based on the key in
+    /// reverse order.
+    fn into_comparator(self) -> Arc<dyn OrderComparator>
+    where
+        Self: Sized,
+    {
+        Arc::new(
+            move |a: &order::Order, b: &order::Order, tokens: &Tokens, solver: &eth::H160| {
+                self.key(a, tokens, solver)
+                    .cmp(&self.key(b, tokens, solver))
+                    .reverse()
+            },
+        )
+    }
+}
+
+pub struct OrderClass;
+impl OrderingKey for OrderClass {
+    type Key = i32;
+
+    // Market orders are preferred over limit orders, as the expectation is that
```

Comment:
nit: should probably be a doc comment on the struct itself. Also would be nice if all strategies had small comment explaining what they prefer and more importantly why this ordering makes sense.

Reply:
Added docs. Let me know if anything is still missing.